safekeeper/src/receive_wal.rs

```diff
@@ -2,216 +2,289 @@
+}
 
-        // Notify the libpq client that it's allowed to send `CopyData` messages
-        self.pg_backend
-            .write_message(&BeMessage::CopyBothResponse)?;
-
-        let r = self
-            .pg_backend
-            .take_stream_in()
-            .ok_or_else(|| anyhow!("failed to take read stream from pgbackend"))?;
-        let mut poll_reader = ProposerPollStream::new(r)?;
-
-        // Receive information about server
-        let next_msg = poll_reader.recv_msg()?;
-        let tli = match next_msg {
-            ProposerAcceptorMessage::Greeting(ref greeting) => {
-                info!(
-                    "start handshake with walproposer {} sysid {} timeline {}",
-                    self.peer_addr, greeting.system_id, greeting.tli,
-                );
-                let server_info = ServerInfo {
-                    pg_version: greeting.pg_version,
-                    system_id: greeting.system_id,
-                    wal_seg_size: greeting.wal_seg_size,
-                };
-                GlobalTimelines::create(spg.ttid, server_info, Lsn::INVALID, Lsn::INVALID)?
-            }
-            _ => {
-                return Err(QueryError::Other(anyhow::anyhow!(
-                    "unexpected message {next_msg:?} instead of greeting"
-                )))
-            }
-        };
+/// Read messages from socket and pass it to WalAcceptor thread. Returns Ok(())
+/// if msg_tx closed; it must mean WalAcceptor terminated, joining it should
+/// tell the error.
+async fn read_network(
+    ttid: TenantTimelineId,
+    pgb_reader: &mut PostgresBackendReader,
+    peer_addr: SocketAddr,
+    msg_tx: Sender<ProposerAcceptorMessage>,
+    // WalAcceptor is spawned when we learn server info from walproposer and
+    // create timeline; handle is put here.
+    acceptor_handle: &mut Option<JoinHandle<anyhow::Result<()>>>,
+    msg_rx: Receiver<ProposerAcceptorMessage>,
+    reply_tx: Sender<AcceptorProposerMessage>,
+) -> Result<(), QueryError> {
+    // Receive information about server to create timeline, if not yet.
+    let next_msg = read_message(pgb_reader).await?;
+    let tli = match next_msg {
+        ProposerAcceptorMessage::Greeting(ref greeting) => {
+            info!(
+                "start handshake with walproposer {} sysid {} timeline {}",
+                peer_addr, greeting.system_id, greeting.tli,
+            );
+            let server_info = ServerInfo {
+                pg_version: greeting.pg_version,
+                system_id: greeting.system_id,
+                wal_seg_size: greeting.wal_seg_size,
+            };
+            GlobalTimelines::create(ttid, server_info, Lsn::INVALID, Lsn::INVALID).await?
+        }
+        _ => {
+            return Err(QueryError::Other(anyhow::anyhow!(
+                "unexpected message {next_msg:?} instead of greeting"
+            )))
+        }
+    };
+    // Now that we have timeline, register the connection.
+    tli.on_compute_connect().await?;
```

Comment:
Maybe it's better to call `on_compute_connect` and `on_compute_disconnect` inside `WalAcceptor::run`? Because currently `on_compute_...` is not synchronized with `tli.process_msg()` calls as they are processed in separate threads.

Reply:
It is actually synchronized, as WalAcceptor does nothing until first message is received, which happens after `on_compute_connect`. I'd like to register connections in global map (to e.g. be able to fetch them with GET), and so place of registration should know peer SocketAddr, in that sense networking thread is closer. But this is not really important, either thread is ok.

Ideally I'd prefer to do that directly in `handle_start_wal_push`, but without explicit timeline creation we don't have a timeline until first message arrives, so I just register as soon as it becomes possible.

Reply:
`on_compute_disconnect` can be called before we finish processing all messages inside `WalAcceptor::run`. That's not a big deal, but it can be confusing if we will track this.

`WalAcceptor` right now looks like a nice connection object, it has I/O channels, no messages are processed until it's spawned and no messages are processed after it was dropped. It seems logical to call `on_compute_connect` in the constructor and `on_compute_disconnect` in the destructor, that's my main point.

Reply:
Haven't thought of that, agree.

Reply:
done it that way